portablemc/optifine.py

```diff
@@ -0,0 +1,659 @@
+                           [of.edition for of in available_of_vers[mcver]])
+                loader = loaders[0]
+            elif re.match(pattern = r"^\d+\.\d+(\.\d+)?:[a-zA-Z0-9_]+$", string = self.version):
+                # when the version is of the form {minecraft version}:{optifine edition}
+                mcver, loader = self.version.split(":")
+                if loader == "latest":
+                    loader = available_of_vers[mcver][0].edition
+                elif loader == "recommended": # takes the latest non-preview loader, if available, or fallback to a preview
+                    loaders = ([of.edition for of in available_of_vers[mcver] if not of.preview] or
+                               [of.edition for of in available_of_vers[mcver]])
+                    loader=loaders[0]
+            elif re.match(pattern  = r"^\d+\.\d+\.\d+-OptiFine_HD_U_[A-Z][0-9](_pre\d)?$", string = self.version):
+                mcver, loader = self.version.split("-OptiFine_") # the version is already in the correct format
+            elif re.match(pattern = r"^(preview_)?OptiFine_\d+\.\d+\.\d+_[A-Z0-9_]+", string=self.version):
+                # in this case, the user directly chosen a filename on the optifine website
+                split_filename = self.version.split("_")
+                if split_filename[0] == "preview":
+                    split_filename.pop(0)
+                mcver = split_filename[1]
+                loader = "_".join(split_filename[2:])
+            else:
+                # no regex matched
+                raise VersionNotFoundError(self.version)
+        except Exception:
+            raise VersionNotFoundError(self.version)
+        if (not mcver in available_of_vers.keys() or
+            not any(loader==entry.edition for entry in available_of_vers[mcver])):
+            raise VersionNotFoundError(self.version)
+        self.version = mcver+"-OptiFine_"+loader
+    def mcver(self):
+        return self.version.split("-OptiFine_")[0]
+    def loader(self):
+        return self.version.split("-OptiFine_")[1]
+    def dl_url(self):
+        edition=self.version.split("-OptiFine_")[1]
+        mcver=self.version.split("-OptiFine_")[0]
+        filename=""
+        if re.match(r"pre\d",edition.split("_")[-1]):
+            filename+="preview_"
+        filename+="OptiFine_"+mcver+"_"+edition
+        return f"http://optifine.net/download?f={filename}.jar"
+
+    def _load_version(self, version: VersionHandle, watcher: Watcher) -> bool:
+        """Same as in ForgeVersion"""
+        if version.id == self.version:
+            return version.read_metadata_file()
+        else:
+            return super()._load_version(version, watcher)
+
+    def _resolve_jar(self, watcher: Watcher) -> None:
+        """Resolves the Optifine installer jar and the client jar"""
+        super()._resolve_jar(watcher)
+        if not (self.context.versions_dir / self.version / "ofInstaller.jar").exists():
```

Comment:
For computing the path of the installer, please prefer `self._hierarchy[0].dir / "ofInstaller.jar"` instead of recomputing the version, it should work properly, replace this everywhere when relevant.

Reply:
Also, you use this "ofInstaller.jar" thing many times, you could define a constant like `INSTALLER_FILE_NAME = "of-installer.jar` (I prefer the kebab-case)